## Evaluating Allocation Funds

### Objective: Grow investment portfolio and find mutual funds with the best returns, lowest fees while reducing risk. 

It's important to consider returns and fees when selecting mutual funds because when we buy/sell. There will be some fees that we have to pay. It has been good practice to understand where that money goes and how much are we willing to pay when we sell our shares.

The scope of this report will cover mutual funds who are:
 - Actively Managed
 - Have 4 or 5 Stars in the Morningstar Rating
 - Categories such as:
     - Allocation Funds (Balanced Funds)

We are going to start with **Allocation Funds** (Balanced Funds).

- Allocation funds, also known as balanced funds, invest in a mix of asset classes, typically including stocks, bonds, and cash.

- Fund Purpose: To provide a diversified portfolio in a single investment.

- Risk/Return Profile: Varies based on the allocation strategy; can range from conservative (more bonds) to aggressive (more stocks).

### Data Dictionary:

Here is a dictionary with descriptions for each column in your dataset:

1. **Symbol**: The unique ticker symbol used to identify the mutual fund.

2. **Description**: A brief description of the mutual fund, including its primary investment strategy or objective.

3. **Product Type**: The category of the financial product, such as mutual fund, ETF, etc.

4. **OneSource List**: Indicates whether the fund is part of the Schwab OneSource list, which typically includes funds that are available without transaction fees.

5. **Transaction Fee**: The fee charged when buying or selling shares of the mutual fund.

6. **Select List**: Indicates if the fund is part of the Schwab Select List, a list of funds selected by Schwab based on specific criteria such as performance and fees.

7. **Schwab Social Funds List**: Indicates if the fund is part of Schwab's socially responsible investment (SRI) list, which includes funds that meet certain environmental, social, and governance (ESG) criteria.

8. **Net Expense Ratio**: The annual percentage fee that the fund charges its shareholders, net of any fee waivers or reimbursements.

9. **Gross Expense Ratio**: The annual percentage fee that the fund charges its shareholders, before any fee waivers or reimbursements.

10. **Inception Date**: The date on which the mutual fund was established.

11. **Monthly NAV Return (1 year)**: The fund's monthly return based on Net Asset Value (NAV) over the past year.

12. **Monthly NAV Return (5 year)**: The fund's monthly return based on NAV over the past five years.

13. **Monthly NAV Return (10 year)**: The fund's monthly return based on NAV over the past ten years.

14. **Monthly NAV Return (Since Inception)**: The fund's monthly return based on NAV since its inception.

15. **Monthly Market Return (1 year)**: The fund's monthly return based on market price over the past year.

16. **Monthly Market Return (5 year)**: The fund's monthly return based on market price over the past five years.

17. **Monthly Market Return (10 year)**: The fund's monthly return based on market price over the past ten years.

18. **Monthly Market Return (Since Inception)**: The fund's monthly return based on market price since its inception.

19. **Monthly Performance as of**: The date as of which the monthly performance metrics are reported.

20. **Quarterly NAV Return (1 year)**: The fund's quarterly return based on NAV over the past year.

21. **Quarterly NAV Return (5 year)**: The fund's quarterly return based on NAV over the past five years.

22. **Quarterly NAV Return (10 year)**: The fund's quarterly return based on NAV over the past ten years.

23. **Quarterly NAV Return (Since Inception)**: The fund's quarterly return based on NAV since its inception.

24. **Quarterly Market Return (1 year)**: The fund's quarterly return based on market price over the past year.

25. **Quarterly Market Return (5 year)**: The fund's quarterly return based on market price over the past five years.

26. **Quarterly Market Return (10 year)**: The fund's quarterly return based on market price over the past ten years.

27. **Quarterly Market Return (Since Inception)**: The fund's quarterly return based on market price since its inception.

28. **Quarterly Performance as of**: The date as of which the quarterly performance metrics are reported.

29. **Morningstar Category**: The category assigned to the fund by Morningstar based on its investment strategy and holdings.

30. **Morningstar Overall**: The overall rating assigned to the fund by Morningstar, based on its risk-adjusted performance.

31. **Morningstar 3 Year**: The three-year rating assigned to the fund by Morningstar, based on its risk-adjusted performance over the past three years.

32. **Morningstar 5 Year**: The five-year rating assigned to the fund by Morningstar, based on its risk-adjusted performance over the past five years.

33. **Morningstar 10 Year**: The ten-year rating assigned to the fund by Morningstar, based on its risk-adjusted performance over the past ten years.

34. **Ranking as of**: The date as of which the Morningstar rankings are reported.

This dictionary provides a clear understanding of each column in your dataset, which will help you analyze and interpret the data effectively.

In [41]:
#imported libraries:
import pandas as pd
import numpy as np
from sklearn.impute import KNNImputer

In [53]:
df = pd.read_csv("cleaned_imputed_knn_dataset.csv")


We are going to start our analysis with **Risk-Adjusted Return Metrics** 

**Sharpe Ratio:** This measures the fund’s excess return per unit of risk. It’s calculated by subtracting the risk-free rate from the fund’s return and then dividing by the standard deviation of the return. Higher Sharpe ratios indicate better risk-adjusted performance.

**Sortino Ratio:** Similar to the Sharpe ratio, but it only considers downside risk (negative returns). This can provide a more accurate measure of a fund’s performance relative to its risk of loss.

**Treynor Ratio:** This measures returns earned in excess of that which could have been earned on a risk-free investment per unit of market risk. It’s particularly useful for funds that are part of a diversified portfolio.

## Questions 

Overall Questions:
- What are the allocation funds that provide the highest monthly returns? 
- What are the allocation funds with the highest monthly market return and Net expense fee equal to zero?

Questions for the fund:
- What is the average Monthly market return for years 1, 5, 10 and since inception.
- What is the median value of Monthly market return for 1, 5, 10 and since inception.
- What is the IQR for the Monthly market return for 1, 5, 10 and since inception.



## Sharpe Ratio:

- This measures the fund’s excess return per unit of risk. It’s calculated by subtracting the risk-free rate from the fund’s return and then dividing by the standard deviation of the return. 
- Higher Sharpe ratios indicate better risk-adjusted performance.
- What are the variables that need to be created to make sure this Sharpe Ratio gets created?


In [43]:
# Overall Questions:
# - What are the allocation funds that provide the highest monthly returns?
# Sort the DataFrame by 'Monthly_NAV_Return_(1_year)' in descending order
sorted_df = df.sort_values(by='Monthly_NAV_Return_(1_year)', ascending=False)
# Select the columns 'Symbol', 'Description', 'Net_Expense_Ratio', 'Monthly_NAV_Return_(1_year)'
highest_monthly_return = sorted_df[['Symbol', 
                                    'Net_Expense_Ratio', 
                                    'Gross_Expense_Ratio', 
                                    'Monthly_NAV_Return_(1_year)',
                                    'Monthly_NAV_Return_(5_year)',
                                    'Monthly_NAV_Return_(10_year)']]
highest_monthly_return.head()


Symbol  Net_Expense_Ratio  Gross_Expense_Ratio  \
275  QSTFX               0.02                 0.02   
258  WWWEX               0.01                 0.02   
212  EKBAX               0.01                 0.01   
188  RPFCX               0.01                 0.01   
261  PLBBX               0.01                 0.02   

     Monthly_NAV_Return_(1_year)  Monthly_NAV_Return_(5_year)  \
275                         0.49                         0.17   
258                         0.39                         0.14   
212                         0.30                         0.11   
188                         0.25                         0.09   
261                         0.24                         0.07   

     Monthly_NAV_Return_(10_year)  
275                         0.086  
258                         0.080  
212                         0.110  
188                         0.060  
261                         0.080

In [44]:
# - What are the highest monthly market return with Net expense equal to zero?
# Filter the DataFrame where Net_Expense_Ratio is equal to zero
filtered_df = df[df['Net_Expense_Ratio'] == 0]

# Sort the filtered DataFrame by Monthly_NAV_Return_(1_year) in descending order
sorted_filtered_df = filtered_df.sort_values(by='Monthly_NAV_Return_(1_year)', ascending=False)

# Select the columns of interest
highest_monthly_return = sorted_filtered_df[['Symbol', 
                                             'Description', 
                                             'Net_Expense_Ratio', 
                                             'Gross_Expense_Ratio', 
                                             'Monthly_NAV_Return_(1_year)',
                                             'Monthly_NAV_Return_(5_year)',
                                             'Monthly_NAV_Return_(10_year)']]

# Display the top results
highest_monthly_return.head()

Symbol                                        Description  \
19  LIVAX  BlackRock LifePath® Index 2055 Fund Investor A...   
8   SWYJX                      Schwab Target 2055 Index Fund   
17  LIPAX  BlackRock LifePath® Index 2050 Fund Investor A...   
11  SWYNX                      Schwab Target 2060 Index Fund   
10  SWYMX                      Schwab Target 2050 Index Fund   

    Net_Expense_Ratio  Gross_Expense_Ratio  Monthly_NAV_Return_(1_year)  \
19                0.0                  0.0                         0.16   
8                 0.0                  0.0                         0.15   
17                0.0                  0.0                         0.15   
11                0.0                  0.0                         0.15   
10                0.0                  0.0                         0.15   

    Monthly_NAV_Return_(5_year)  Monthly_NAV_Return_(10_year)  
19                         0.09                         0.080  
8                          0.09                         0.080  
17                         0.09                         0.080  
11                         0.09                         0.080  
10                         0.08                         0.078

In [55]:
# - What are the highest monthly market return with Gross Expense equal to zero?
# Filter the DataFrame where Gross_Expense_Ratio is equal to zero
filtered_df = df[df['Gross_Expense_Ratio'] == 0]

# Sort the filtered DataFrame by Monthly_Market_Return_(1_year) in descending order
sorted_filtered_df = filtered_df.sort_values(by='Monthly_NAV_Return_(1_year)', ascending=False)

# Select the columns of interest
highest_monthly_return = sorted_filtered_df[['Symbol', 
                                             'Description', 
                                             'Net_Expense_Ratio', 
                                             'Gross_Expense_Ratio', 
                                             'Monthly_NAV_Return_(1_year)',
                                             'Monthly_NAV_Return_(5_year)',
                                             'Monthly_NAV_Return_(10_year)']]

highest_monthly_return.head()

# # Questions for the fund:
# # - What is the average Monthly NAV return for 1, 5, 10 and since inception.
# averages_Monthly_since_inception = df[['Symbol', 
#                                              'Description', 
#                                              'Net_Expense_Ratio', 
#                                              'Gross_Expense_Ratio', 
#                                              'Monthly_NAV_Return_(1_year)',
#                                              'Monthly_NAV_Return_(5_year)',
#                                              'Monthly_NAV_Return_(10_year)',
#                                              'Inception_Date']]

# # - What is the IQR for the Monthly market return for 1, 5, 10 and since inception.


Symbol                                        Description  \
19  LIVAX  BlackRock LifePath® Index 2055 Fund Investor A...   
8   SWYJX                      Schwab Target 2055 Index Fund   
17  LIPAX  BlackRock LifePath® Index 2050 Fund Investor A...   
11  SWYNX                      Schwab Target 2060 Index Fund   
10  SWYMX                      Schwab Target 2050 Index Fund   

    Net_Expense_Ratio  Gross_Expense_Ratio  Monthly_NAV_Return_(1_year)  \
19                0.0                  0.0                         0.16   
8                 0.0                  0.0                         0.15   
17                0.0                  0.0                         0.15   
11                0.0                  0.0                         0.15   
10                0.0                  0.0                         0.15   

    Monthly_NAV_Return_(5_year)  Monthly_NAV_Return_(10_year)  
19                         0.09                         0.080  
8                          0.09                         0.080  
17                         0.09                         0.080  
11                         0.09                         0.080  
10                         0.08                         0.078

In [46]:
# Select the columns of interest for calculating averages
average_columns = ['Monthly_NAV_Return_(1_year)', 
                   'Monthly_NAV_Return_(5_year)', 
                   'Monthly_NAV_Return_(10_year)', 
                   'Quarterly_NAV_Return_(1_year)',
                   'Quarterly_NAV_Return_(5_year)',
                   'Quarterly_NAV_Return_(10_year)',
                   'Quarterly_NAV_Return_(Since_Inception)']

# Calculate the average for each of the selected columns
average_nav_returns = df[average_columns].mean()

# Display the average NAV returns
average_nav_returns


Monthly_NAV_Return_(1_year)               0.109161
Monthly_NAV_Return_(5_year)               0.064280
Monthly_NAV_Return_(10_year)              0.060783
Quarterly_NAV_Return_(1_year)             0.155629
Quarterly_NAV_Return_(5_year)             0.075706
Quarterly_NAV_Return_(10_year)            0.063958
Quarterly_NAV_Return_(Since_Inception)    0.070455
dtype: float64

In [47]:
import pandas as pd

# Select the columns of interest
columns = ['Monthly_NAV_Return_(1_year)', 
           'Monthly_NAV_Return_(5_year)', 
           'Monthly_NAV_Return_(10_year)', 
           'Quarterly_NAV_Return_(1_year)',
           'Quarterly_NAV_Return_(5_year)',
           'Quarterly_NAV_Return_(10_year)',
           'Quarterly_NAV_Return_(Since_Inception)']

# Create an empty dictionary to store the results
stats = {}

# Calculate the required statistics for each column
for column in columns:
    data = df[column].dropna()  # Drop NaN values if any
    stats[column] = {
        'Min': data.min(),
        'Q1': data.quantile(0.25),
        'Median': data.median(),
        'Average': data.mean(),
        'Q3': data.quantile(0.75),
        'Max': data.max(),
        'Range': data.max() - data.min(),
        'Standard Deviation': data.std(),
        'Variance': data.var()
    }

# Convert the results to a DataFrame for better readability
stats_df = pd.DataFrame(stats).T

# # Display the statistics
# print(stats_df)
stats_df


Min    Q1  Median   Average    Q3  \
Monthly_NAV_Return_(1_year)             0.00  0.07   0.100  0.109161  0.14   
Monthly_NAV_Return_(5_year)             0.02  0.05   0.068  0.064280  0.08   
Monthly_NAV_Return_(10_year)            0.02  0.05   0.060  0.060783  0.07   
Quarterly_NAV_Return_(1_year)           0.03  0.11   0.150  0.155629  0.20   
Quarterly_NAV_Return_(5_year)           0.03  0.06   0.080  0.075706  0.09   
Quarterly_NAV_Return_(10_year)          0.03  0.05   0.062  0.063958  0.08   
Quarterly_NAV_Return_(Since_Inception)  0.02  0.06   0.070  0.070455  0.08   

                                         Max  Range  Standard Deviation  \
Monthly_NAV_Return_(1_year)             0.49   0.49            0.054721   
Monthly_NAV_Return_(5_year)             0.17   0.15            0.020479   
Monthly_NAV_Return_(10_year)            0.11   0.09            0.017384   
Quarterly_NAV_Return_(1_year)           0.64   0.61            0.064602   
Quarterly_NAV_Return_(5_year)           0.21   0.18            0.023351   
Quarterly_NAV_Return_(10_year)          0.11   0.08            0.017344   
Quarterly_NAV_Return_(Since_Inception)  0.17   0.15            0.020506   

                                        Variance  
Monthly_NAV_Return_(1_year)             0.002994  
Monthly_NAV_Return_(5_year)             0.000419  
Monthly_NAV_Return_(10_year)            0.000302  
Quarterly_NAV_Return_(1_year)           0.004173  
Quarterly_NAV_Return_(5_year)           0.000545  
Quarterly_NAV_Return_(10_year)          0.000301  
Quarterly_NAV_Return_(Since_Inception)  0.000420

## Sharpe Ratio: 
This measures the fund’s excess return per unit of risk. It’s calculated by subtracting the risk-free rate from the fund’s return and then dividing by the standard deviation of the return. Higher Sharpe ratios indicate better risk-adjusted performance.

In [50]:
# Select the columns of interest for calculating averages
columns = ['Monthly_NAV_Return_(1_year)', 
                   'Quarterly_NAV_Return_(1_year)',
                   'Quarterly_NAV_Return_(Since_Inception)']

# Calculate the average for each of the selected columns
monthly_NAV_Return_one_year_mean = df['Monthly_NAV_Return_(1_year)'].mean()
#print('This is the average for NAV Monthly',monthly_NAV_Return_one_year_mean)

monthly_NAV_Return_one_year_std = df['Monthly_NAV_Return_(1_year)'].std()
#print('This is the Standard Dev for NAV Monthly',monthly_NAV_Return_one_year_std)
risk_free_rate = 0.04224

df['sharpe_ratio'] = (df['Monthly_NAV_Return_(1_year)'] - risk_free_rate) /monthly_NAV_Return_one_year_std

sorted_df = df.sort_values(by='sharpe_ratio', ascending=False)
sharpe_ratio_return = sorted_df[['Symbol', 
                                 'sharpe_ratio',
                                    'Net_Expense_Ratio', 
                                    'Gross_Expense_Ratio', 
                                    'Monthly_NAV_Return_(1_year)',
                                    'Monthly_NAV_Return_(5_year)',
                                    'Monthly_NAV_Return_(10_year)']]
sharpe_ratio_return[0:20]
 

Symbol  sharpe_ratio  Net_Expense_Ratio  Gross_Expense_Ratio  \
275  QSTFX      8.182608               0.02                 0.02   
258  WWWEX      6.355154               0.01                 0.02   
212  EKBAX      4.710446               0.01                 0.01   
188  RPFCX      3.796719               0.01                 0.01   
261  PLBBX      3.613973               0.01                 0.02   
113  GWPFX      3.431228               0.01                 0.01   
219  VALIX      3.248482               0.01                 0.01   
269  HCMEX      3.065737               0.02                 0.02   
190  FTCOX      3.065737               0.01                 0.01   
270  QMLFX      3.065737               0.02                 0.02   
202  PAEAX      2.882992               0.01                 0.01   
246  CVLOX      2.700246               0.01                 0.01   
245  IAAAX      2.517501               0.01                 0.01   
247  EAEAX      2.517501               0.01                 0.02   
274  HASAX      2.517501               0.02                 0.02   
157  TSALX      2.517501               0.01                 0.01   
181  SSIAX      2.517501               0.01                 0.01   
27   TRRLX      2.334755               0.01                 0.01   
205  CVTRX      2.334755               0.01                 0.01   
107  FAJTX      2.334755               0.01                 0.01   

     Monthly_NAV_Return_(1_year)  Monthly_NAV_Return_(5_year)  \
275                         0.49                         0.17   
258                         0.39                         0.14   
212                         0.30                         0.11   
188                         0.25                         0.09   
261                         0.24                         0.07   
113                         0.23                         0.10   
219                         0.22                         0.08   
269                         0.21                         0.09   
190                         0.21                         0.09   
270                         0.21                         0.09   
202                         0.20                         0.08   
246                         0.19                         0.10   
245                         0.18                         0.08   
247                         0.18                         0.09   
274                         0.18                         0.08   
157                         0.18                         0.09   
181                         0.18                         0.09   
27                          0.17                         0.09   
205                         0.17                         0.11   
107                         0.17                         0.09   

     Monthly_NAV_Return_(10_year)  
275                         0.086  
258                         0.080  
212                         0.110  
188                         0.060  
261                         0.080  
113                         0.100  
219                         0.090  
269                         0.086  
190                         0.080  
270                         0.080  
202                         0.080  
246                         0.070  
245                         0.080  
247                         0.090  
274                         0.074  
157                         0.090  
181                         0.080  
27                          0.086  
205                         0.100  
107                         0.090

## Let's look at the net expense ratio at zero percent.

In [49]:
filtered_df = df[df['Net_Expense_Ratio'] == 0]

# Sort the filtered DataFrame by Sharpe Ratio in descending order
sorted_filtered_df = filtered_df.sort_values(by='sharpe_ratio', ascending=False)

# Select the columns of interest
zero_net_expense_with_high_sharpe_ratio = sorted_filtered_df[['Symbol',
                                             'sharpe_ratio', 
                                             'Monthly_NAV_Return_(1_year)',
                                             'Monthly_NAV_Return_(5_year)',
                                             'Monthly_NAV_Return_(10_year)']]

zero_net_expense_with_high_sharpe_ratio[0:20]

Symbol  sharpe_ratio  Monthly_NAV_Return_(1_year)  \
19  LIVAX      2.152010                         0.16   
8   SWYJX      1.969264                         0.15   
17  LIPAX      1.969264                         0.15   
11  SWYNX      1.969264                         0.15   
10  SWYMX      1.969264                         0.15   
3   SWYOX      1.969264                         0.15   
7   SWYHX      1.786519                         0.14   
14  LIHAX      1.786519                         0.14   
6   SWYGX      1.603774                         0.13   
41  SWIRX      1.421028                         0.12   
2   SWYFX      1.421028                         0.12   
16  LIKAX      1.421028                         0.12   
37  SWDRX      1.055537                         0.10   
15  LIJAX      1.055537                         0.10   
1   SWYEX      1.055537                         0.10   
42  TRRAX      0.872792                         0.09   
20  SWBGX      0.872792                         0.09   
43  TRRFX      0.690047                         0.08   
36  SWHRX      0.690047                         0.08   
21  TRRUX      0.690047                         0.08   

    Monthly_NAV_Return_(5_year)  Monthly_NAV_Return_(10_year)  
19                        0.090                         0.080  
8                         0.090                         0.080  
17                        0.090                         0.080  
11                        0.090                         0.080  
10                        0.080                         0.078  
3                         0.064                         0.064  
7                         0.080                         0.080  
14                        0.080                         0.080  
6                         0.080                         0.072  
41                        0.070                         0.070  
2                         0.070                         0.064  
16                        0.070                         0.070  
37                        0.060                         0.060  
15                        0.060                         0.060  
1                         0.060                         0.064  
42                        0.050                         0.050  
20                        0.060                         0.060  
43                        0.050                         0.050  
36                        0.050                         0.060  
21                        0.050                         0.050

A high Sharpe Ratio, as observed in your top funds, suggests that these investments are effectively balancing risk and return, making them attractive options. However, always consider other factors such as the consistency of returns, management quality, and economic conditions before making investment decisions.

- QSTFX: With a Sharpe Ratio of 8.18, this fund is delivering very high returns compared to its risk. Its expense ratios are low, contributing to its high performance.

- WWWEX: Also shows a strong performance with a Sharpe Ratio of 6.36. Its returns are slightly lower than QSTFX, but it still significantly outperforms in terms of risk-adjusted returns.

- EKBAX, RPFCX, PLBBX: These funds have Sharpe Ratios above 3, indicating excellent performance. They have lower returns compared to QSTFX and WWWEX but are still providing good returns relative to their risk.